Add python packages: pyensembl and its dependencies, gffutils

moritzschaefer opened the pull request:
##### Motivation for this change

Replaces PR https://github.com/NixOS/nixpkgs/pull/87551

Both pyenseml and gffutils are commonly used python packages (> 100 stars on GitHub), so I added them to nix packages. 

###### Things done

- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) (only pyensembl has a supplemental binary)
- [x] Ensured that relevant documentation is up to date
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [-] Mostly Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md). I designed the derivations 1:1 to existing ones.

moritzschaefer committed:
Add pyensembl library (https://github.com/openvax/pyensembl)

moritzschaefer committed:
Add gffutils python package

moritzschaefer committed:
scikit-plot: 0.3.7

moritzschaefer committed:
pybedtools: 0.8.1

moritzschaefer committed:
Add checks for python modules

moritzschaefer committed:
pybedtools: Add missing dependency

jonringer commented on pkgs/development/python-modules/datacache/default.nix:
```diff
@@ -0,0 +1,27 @@
+    sha256 = "b2ca31b2b9d3803a49645ab4f5b30fdd0820e833a81a6952b4ec3a68c8ee24a7";
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks nose mock ];
+
+  checkPhase = ''
+      ${python.interpreter} -c 'import datacache'
+  '';
+
+  doCheck = true;
```
```suggestion
  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks nose mock ];

  pythonImportsCheck = [ "datacache" ];
```

jonringer commented on pkgs/development/python-modules/datacache/default.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with stdenv.lib; {
+    homepage = "https://github.com/openvax/datacache";
+    description = "Helpers for transparently downloading datasets";
+    license = licenses.asl20;
```
please add yourself as a maintainer

jonringer commented on pkgs/development/python-modules/datacache/default.nix:
```diff
@@ -0,0 +1,27 @@
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks nose mock ];
```
not needed at runtime, should go in checkInputs if anything
```suggestion
  propagatedBuildInputs = [ pandas appdirs progressbar33 requests typechecks mock ];
```

jonringer commented on pkgs/development/python-modules/gffutils/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+  checkPhase = ''
+    # sh gffutils/test/data/download-large-annotation-files.sh
+    # nosetests
```
why no test suite?
```suggestion
    # nosetests
```

moritzschaefer replied:
The tests fail inside the nix environment

jonringer replied:
general consensus is to just ignore test cases that fail due to sandbox environment. But some test failures may be indications that the package is in an actual broken state.

tests are also very useful for verifying that dependency bumps don't cause regressions.

jonringer commented on pkgs/development/python-modules/gtfparse/default.nix:
```diff
@@ -0,0 +1,29 @@
+  checkInputs = [ nose six ];
+  propagatedBuildInputs = [ numpy pandas ];
+
+  checkPhase = ''
+    # PYTHONPATH='test' nosetests # fails because six is not found
+    ${python.interpreter} -c 'import gtfparse'
+  '';
```
use pythonImportsCheck

jonringer commented on pkgs/development/python-modules/memoized-property/default.nix:
```diff
@@ -0,0 +1,26 @@
+  checkInputs = [ ];
+  propagatedBuildInputs = [ ];
+
+  checkPhase = ''
+      ${python.interpreter} -c 'import memoized_property'
+  '';
+
```
use pythonImportsCheck

jonringer commented on pkgs/development/python-modules/pybedtools/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ bash, stdenv, buildPythonPackage, fetchPypi, numpy, pandas, pysam, six,
+  pytest, numpydoc, psutil, pyyaml, sphinx, zlib, bedtools, python
+}:
```
```suggestion
}:

```

jonringer commented on pkgs/development/python-modules/pybedtools/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+  checkPhase = ''
+    # pytest -v --doctest-modules
+    # ${python.interpreter} -c 'import pybedtools'  # test and import do not work in checkPhase, because the built pyx file cannot be included
```
pythonImportsCheck

jonringer commented on pkgs/development/python-modules/pyensembl/default.nix:
```diff
@@ -0,0 +1,28 @@
+  checkInputs = [ ];
+  propagatedBuildInputs = [ numpy pandas datacache six memoized-property gtfparse serializable tinytimer serializable ];
+
+  checkPhase = ''
+    ${python.interpreter} -c 'import pyensembl'  # fails only in build environment because pyensembl creates a file in root directory
```
pythonImportsCheck

jonringer commented on pkgs/development/python-modules/scikit-plot/default.nix:
```diff
@@ -0,0 +1,36 @@
+    sha256 = "2c7948817fd2dc06879cfe3c1fdde56a8e71fa5ac626ffbe79f043650baa6242";
+  };
+
+  buildInputs = [  ];
```
```suggestion
```

jonringer commented on pkgs/development/python-modules/scikit-plot/default.nix:
```diff
@@ -0,0 +1,36 @@
+  checkInputs = [ nose ];
+  propagatedBuildInputs = [ matplotlib scikitlearn scipy joblib ];
+
+  doCheck = true;
+
```
default for `buildPythonPackage`
```suggestion;
```

jonringer commented on pkgs/development/python-modules/serializable/default.nix:
```diff
@@ -0,0 +1,27 @@
+{ stdenv, buildPythonPackage, fetchPypi,
+typechecks, six, simplejson, nose}:
```
```suggestion
typechecks, six, simplejson, nose}:

```

jonringer commented on pkgs/development/python-modules/serializable/default.nix:
```diff
@@ -0,0 +1,27 @@
+    nosetests
+  '';
+
+  # Tests require extra dependencies
+  doCheck = true;
+
```
default
```suggestion
```

jonringer commented on pkgs/development/python-modules/tinytimer/default.nix:
```diff
@@ -0,0 +1,27 @@
+    sha256 = "6ad13c8f01ab6094e58081a5367ffc4c5831f2d6b29034d2434d8ae106308fa5";
+  };
+
+  checkInputs = [ ];
+  propagatedBuildInputs = [ ];
+
+  checkPhase = ''
+    ${python.interpreter} -c 'import tinytimer'
+  '';
+
+  # Tests require extra dependencies
+  doCheck = true;
+
```
```suggestion
  pythonImportsCheck = [ "tinytimer" ];
```

jonringer commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7405,6 +7405,25 @@ in {
 
   rxv     = callPackage ../development/python-modules/rxv     { };
 
+  typechecks = callPackage ../development/python-modules/typechecks { };
+
+  datacache = callPackage ../development/python-modules/datacache { };
+
+  memoized-property = callPackage ../development/python-modules/memoized-property { };
+
+  gtfparse = callPackage ../development/python-modules/gtfparse { };
+
+  serializable = callPackage ../development/python-modules/serializable { };
+
+  tinytimer = callPackage ../development/python-modules/tinytimer { };
+
+  pyensembl = callPackage ../development/python-modules/pyensembl { };
+
+  gffutils = callPackage ../development/python-modules/gffutils { };
+
+  scikit-plot = callPackage ../development/python-modules/scikit-plot { };
+
+  pybedtools = callPackage ../development/python-modules/pybedtools { };
```
please attempt to sort these

jonringer replied:
it has been evolving over time... some people grouped similar packages, some people just dropped them anywhere. At the very least, random listings should be avoided

jonringer reviewed (commented):
please add yourself as a maintainer for all packages that you're adding.

please attempt to sort the package entries in `python-packages`

moritzschaefer committed:
Add myself as maintainer to added packages; Declutter new nix files

moritzschaefer commented:
Thank you for your comments @jonringer! I have simplified all the files and added myself as maintainer. Please let me know if something still requires improvements.

moritzschaefer committed:
sort new packages alphabetically

moritzschaefer committed:
move scikit-plot to other scikit packages in python-packages.nix

moritzschaefer committed:
Fix typos

moritzschaefer commented:
@jonringer After having applied all the requested changes, I wanted to da a final rebase on master so I did

> git checkout python-packages
> git pull --rebase official master  # official points to github.com/NIXOS/nixpgs.git

This resulted in a merge commit and now the PR is blown up with 370 commits. Is this an issue and if yes, would you mind telling me what went wrong?

jonringer commented:
with --rebase, there shouldn't be a merge commit.

And it looks like you pulled in master. Assuming the merge is the main culprit, you could do `git reset HEAD^ --hard` but, I'm not sure exactly what you've done.

If all else failes, you could cherry pick the commits into a new branch and just push that

moritzschaefer commented:
Indeed trashing the mere commit seemed to have fixed it. So no cherry-picking-in-new-branch required. 
Would you approve to merge like this?

moritzschaefer committed:
Disable failing tests

moritzschaefer committed:
disable failing input test

moritzschaefer committed:
Merge branch 'master' into python-packages

jonringer commented:
if you still want to get this into nixpkgs, you can break this up into smaller PRs, just adding a package at a time. It will be a lot less time intensive than having to cover all of it.

moritzschaefer commented:
Dear @jonringer,

thanks for the response. I'll transition to adding these packages into my overlays and close this PR as there are definitely more important PRs (from me and others) for the moment.

Still, I am kind of missing some kind of strategy for pypi packages. In a conversation on the nixos-discourse, I got the notion that it is not desired to add the majority of pypi packages to nixpkgs. However it is really cumbersome to build a derivation for each single python package, when you're working in a non-mainstream field. 

Is there some kind of vision for this issue?

jonringer commented:
In general the preference is to have python packages available which enable an application, and less so to "just have the package". 



> However it is really cumbersome to build a derivation for each single python package, when you're working in a non-mainstream field.

You can use nix with venv to bring in missing python packages. I got a video about it https://www.youtube.com/watch?v=jXd-hkP4xnU , and some docs here: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#how-to-consume-python-modules-using-pip-in-a-virtual-environment-like-i-am-used-to-on-other-operating-systems

If you still want a pure nix way to do it, I would look at `mach-nix`https://github.com/DavHau/mach-nix
